### PR TITLE
CXP-768 Gate account re-parenting on org/OU sync opt-in

### PR DESCRIPTION
**Pull request by JavierCarnelli-ConductorOne**
accountResourceType.List() re-parented every account onto its Organizations Root/OU whenever organizations:ListParents succeeded, regardless of whether the organization/organizational_unit resource types (OptInRequired, off by default) were actually being synced. In the default orgs+sso configuration this left every account with a parent pointing at a resource that was never synced, and c1 silently drops the reference as a dangling MISSING RESOURCE.

Thread cli.ConnectorOpts through New() (previously discarded) to learn whether this sync run will actually emit organization/ organizational_unit, and gate both the ListParents call and the resulting parent attachment on it -- including the resolved parent's specific type, so partial opt-in can't produce a dangling parent either.

### CODE REVIEW

File: pkg/connector/account.go [hunk 1]
```diff
@@ -129,6 +129,14 @@ type accountResourceType struct {
 	identityInstance *awsSsoAdminTypes.InstanceMetadata
 	identityClient   client.IdentityStoreClient
 	region           string
+
+	// willSyncOrganization/willSyncOrganizationalUnit report whether this sync run will
+	// actually sync the corresponding (OptInRequired) hierarchy resource type. Account
+	// re-parenting (see List) is gated on these so accounts never point at a Root/OU
+	// resource that this run never syncs, which would otherwise leave a dangling
+	// "MISSING RESOURCE" parent. See CXP-768.
```
Comment (Bencheng21) [inline pkg/connector/account.go]: delete CXP-768, because it's public repo

**Review by Bencheng21 — APPROVED** [review]

**Comment by Bencheng21** [comment]
Adjacent edge not addressed: the OU→Root parent edge (organizationalUnitResource → WithParentResourceID) has the same dangling shape if organizational_unit is synced while organization (Root) isn't. In practice the OU crawl is seeded from the Root, so "OU without org" is likely unreachable — but if that invariant changes, the same MISSING RESOURCE could reappear one tier up. Note on the ticket. 

Could you confirm with `OU without org is unreachable?`

**Comment by JavierCarnelli-ConductorOne** [comment]
Hey @Bencheng21 ! About your question:
> Could you confirm with `OU without org is unreachable?`

I can confirm. "OU without org" is unreachable. As you mentioned, this is covered by design. If the Root Org is not synced, no OU will be on the result set.
This is likely not going to change. So we should be good with it

File: pkg/connector/account.go (hunk 1 above)
Comment (JavierCarnelli-ConductorOne) [reply pkg/connector/account.go]: Addressed in [58f3b2b](https://github.com/ConductorOne/baton-aws/pull/139/commits/58f3b2b7a83fc941af1dbc87d883f05ce95e8076)

Thanks!

File: pkg/connector/connector.go [hunk 2]
```diff
@@ -240,9 +246,18 @@ func validateConfig(awsc *cfg.Aws) error {
...
-func New(ctx context.Context, awsc *cfg.Aws, _ *cli.ConnectorOpts) (connectorbuilder.ConnectorBuilderV2, []connectorbuilder.Opt, error) {
+func New(ctx context.Context, awsc *cfg.Aws, connectorOpts *cli.ConnectorOpts) (connectorbuilder.ConnectorBuilderV2, []connectorbuilder.Opt, error) {
 	l := ctxzap.Extract(ctx)
 
+	// Default to "will sync" when the caller gave no explicit resource-type filter (or no
+	// ConnectorOpts at all), matching ConnectorOpts.WillSyncResourceType's own default.
+	willSyncOrganization := true
+	willSyncOrganizationalUnit := true
+	if connectorOpts != nil {
+		willSyncOrganization = connectorOpts.WillSyncResourceType(resourceTypeOrganization.Id)
```
Comment (sergiocorral-conductorone) [inline pkg/connector/connector.go:259]: **[Major — please confirm]** I checked this against the actual baton-sdk source (`pkg/cli/cli.go`): `WillSyncResourceType` returns `true` whenever `SyncFilterIsExplicit()` is false (i.e. `SyncResourceTypeIDs` is empty) — it does **not** check the `OptInRequired` annotation at all. The doc comment on `SyncResourceTypeIDs` itself says an empty filter means "sync everything the connector advertises (subject to OptInRequired{} annotations)" — implying OptInRequired exclusion happens somewhere *outside* this helper. I also checked `pkg/sync/syncer.go` in the SDK: the local/direct syncer only filters resource types when `len(s.syncResourceTypes) > 0`; with an empty filter it syncs everything the connector advertises, OptInRequired or not.

So `willSyncOrganization`/`willSyncOrganizationalUnit` will compute to `true` whenever this connector process runs with no *explicit* `--sync-resource-types` filter — which is exactly the "default orgs+sso configuration" this PR's description says is the bug scenario. If that's how a real customer sync actually reaches this connector (no explicit filter, relying on OptInRequired being enforced elsewhere), this PR reintroduces the dangling parent it's meant to fix, just via a different code path.

This only doesn't matter if the ConductorOne platform *always* sends an explicit `SyncResourceTypeIDs` list (derived from the org's opt-in selections) when it triggers a real sync job, making `SyncFilterIsExplicit()` true and the gate correct — that's platform-internal behavior neither I nor the second-model verifier could confirm from the SDK/PR alone. Could you confirm which path applies for a real customer sync? If there's any path where the filter can be empty/non-explicit for a production sync, consider gating on `connectorOpts.SyncFilterIsExplicit() && connectorOpts.WillSyncResourceType(id)` instead, and add a test at the `New()` boundary covering the default (no filter) case.

File: pkg/connector/account.go [hunk 3]
```diff
@@ -190,15 +198,33 @@ func (o *accountResourceType) List(ctx context.Context, _ *v2.ResourceId, opts r
...
-			resourceOpts = append(resourceOpts, resourceSdk.WithParentResourceID(parentID))
+		//
+		// Gated on willSyncOrganization/willSyncOrganizationalUnit: organization and
+		// organizational_unit are OptInRequired, so a sync run may never emit them. Pointing an
+		// account's parent at a resource type this run isn't syncing produces a dangling
+		// "MISSING RESOURCE" parent that c1 silently drops (see permissionSetRoleID's comment on
+		// dangling references). Skip the ListParents call entirely when neither hierarchy type
+		// will be synced; when only one is, still resolve the parent but only attach it if its
+		// resolved type is one this run will actually sync.
+		if o.willSyncOrganization || o.willSyncOrganizationalUnit {
```
Comment (sergiocorral-conductorone) [inline pkg/connector/account.go]: **[Minor]** The outer guard `if o.willSyncOrganization || o.willSyncOrganizationalUnit` still triggers the rate-limited `ListParents` call when org=false and OU=true. But `organizationResourceType` declares `ChildResourceType: organizational_unit` (checked in `organization.go`) — OU is only ever crawled as a child of the org resource, so if org isn't synced, no OU is ever emitted regardless of `willSyncOrganizationalUnit`. Both attach branches below already require `willSyncOrganization` for exactly this reason. In the (org=false, OU=true) config, this call can never produce an attachment — it's a wasted per-account AWS Organizations API call. Simplifying the guard to `if o.willSyncOrganization` would avoid the waste and make the org→OU dependency explicit.

**Review by agustin-conductor — APPROVED** [review]

**Review by mateoHernandez123 — APPROVED** [review]

File: pkg/connector/account.go (hunk 3 above)
Comment (FeliLucero1) [inline pkg/connector/account.go]: Yeah, seconding @sergiocorral-conductorone's point above — hit the same thing independently running a deep-code-review pass on this branch.

Confirmed it algebraically too: `willSyncParentType` is `(A && x) || (B && y && x)` where `x = willSyncOrganization`. Since `x` is required in both branches anyway, distributing gives `x && (A || (B && y))` — same truth table, but it also means the outer guard only ever needs to be `if o.willSyncOrganization`, dropping the OU-only case that (as pointed out above) burns a `ListParents` call per account for nothing. Concretely:

```go
if o.willSyncOrganization {
    parentID, accessDenied, err := accountParentResourceID(ctx, o.orgClient, accountId)
    if err != nil {
        return nil, nil, err
    }
    if accessDenied {
        orgReadDenied = true
    }
    if parentID != nil {
        willSyncParentType := parentID.ResourceType == resourceTypeOrganization.Id ||
            (parentID.ResourceType == resourceTypeOrganizationalUnit.Id && o.willSyncOrganizationalUnit)
        if willSyncParentType {
            resourceOpts = append(resourceOpts, resourceSdk.WithParentResourceID(parentID))
        }
    }
}
```

Fixes the wasted call and makes the org→OU dependency structural instead of only living in a comment. Also worth adding the (org=false, OU=true) case to the test file while you're in here — it's the one combination that isn't covered yet.

File: pkg/connector/account.go [hunk 4]
```diff
@@ -952,15 +978,19 @@ func accountBuilder(
 	identityInstance *awsSsoAdminTypes.InstanceMetadata,
 	region string,
 	identityClient client.IdentityStoreClient,
+	willSyncOrganization bool,
+	willSyncOrganizationalUnit bool,
```
Comment (FeliLucero1) [inline pkg/connector/account.go]: Small nit, unrelated to the thread above: `accountBuilder` is now 8 positional params deep, and the last two (`willSyncOrganization`, `willSyncOrganizationalUnit`) are adjacent untyped bools — nothing stops a future edit from swapping them at a call site and the compiler won't catch it.

Right now the one real wiring call site (`connector.go`, `c.willSyncOrganization, c.willSyncOrganizationalUnit`) is correct, but none of the existing tests would catch a swap there specifically — `organization_test.go` and `permission_set_assignment_behavior_test.go` both call `accountBuilder` directly with local named vars or hardcode `true, true`, bypassing the actual `AWS{}` → `ResourceSyncers()` wiring path.

Not blocking — just flagging. Might be worth either a test that drives that real path with distinct flag values, or grouping the two bools into a tiny named struct so a transposition becomes a compile error instead of a silent behavior change.

**Review by luisina-santos — APPROVED** [review]
Approving — the fix is correct and well-tested. Two follow-ups worth addressing before merge, neither changing the approach.

**Resolving the [Major] on `connector.go:257`** — @sergiocorral-conductorone's reading of the SDK is accurate: in `baton-sdk@v0.18.4`, `WillSyncResourceType` (`pkg/cli/cli.go:89`) returns `true` whenever the filter is empty and never consults `OptInRequired`. Grepping the whole SDK, the only non-test references to `OptInRequired` are that doc comment and `connectorbuilder.go:426`, where it's just reported in capabilities metadata — the syncer never enforces it.

That said, the concern doesn't apply: C1 always sends an explicit `SyncResourceTypeIDs` list when it triggers a sync, so `SyncFilterIsExplicit()` is true and the gate does a real set lookup on a production path.

And the `true` default is *positively* correct, not merely harmless. In an unfiltered run (local CLI), the syncer syncs everything the connector advertises — org and OU included — so the resolved parent **does** get emitted and is not dangling. Defaulting to "will sync" is precisely right there. Gating on `SyncFilterIsExplicit() && WillSyncResourceType(id)` would make that case *worse*: it would suppress re-parenting on a run that actually does sync the hierarchy. Current code is right as-is.

**Please address — [Minor], `account.go:209`** (raised by @sergiocorral-conductorone, seconded by @FeliLucero1). Both `willSyncParentType` branches already require `o.willSyncOrganization`, so the outer `||` guard is wider than the inner condition can ever use: in (org=false, OU=true) it fires a rate-limited `ListParents` per account that can never produce an attachment. Collapsing the outer guard to `if o.willSyncOrganization` removes the wasted call and makes the org→OU dependency structural rather than comment-only. @FeliLucero1's block above is the change verbatim.

While in there: the new tests cover (false,false) and (true,false), but **(false,true) is the one combination not covered** — and it's the case this guard change is about. Worth adding, asserting `listParentsCalls == 0`.

**Non-blocking — [Nit], `account.go:982`** (@FeliLucero1). Agreed and confirmed: `accountBuilder` is 8 positional params with two adjacent untyped bools, and neither `newOrgAccountWithSyncFilter` nor the behavior-test helper exercises the real `AWS{}` → `ResourceSyncers()` wiring, so a transposition at that one live call site would be a silent behavior change rather than a compile error. Fine to defer, but if the tests are being touched for the (false,true) case anyway, a small named struct for the two flags is cheap insurance.

File: pkg/connector/connector.go (hunk 2 above)
Comment (JavierCarnelli-ConductorOne) [reply pkg/connector/connector.go:259]: Traced this through the SDK (v0.19.1): OptInRequired isn't enforced anywhere in pkg/sync/pkg/cli — it's just an annotation. Enforcement only happens through SyncResourceTypeIDs, the same filter this PR already gates on.

That means the "no explicit filter" case is actually fine: organization/organizational_unit are registered whenever
orgs+SSO are enabled, so an empty filter (e.g. CLI with no --sync-resource-types) syncs everything, org/OU included — willSyncOrganization = true is correct there, not a gap.

The real dangling-parent case is an explicit filter that includes account but excludes org/OU (opted-out org)  — already handled by this PR's gate.

File: pkg/connector/account.go (hunk 3 above)
Comment (JavierCarnelli-ConductorOne) [reply pkg/connector/account.go]: Thanks for mentioning this!
Addressed in [8d39453](https://github.com/ConductorOne/baton-aws/pull/139/commits/8d3945361a16a652c360a158d886f4836368e1e0)

Comment (JavierCarnelli-ConductorOne) [reply pkg/connector/account.go]: I agree, addressed in [8d39453](https://github.com/ConductorOne/baton-aws/pull/139/commits/8d3945361a16a652c360a158d886f4836368e1e0)

Thanks!

File: pkg/connector/account.go (hunk 4 above)
Comment (JavierCarnelli-ConductorOne) [reply pkg/connector/account.go]: Addressed in [dbf9087](https://github.com/ConductorOne/baton-aws/pull/139/commits/dbf9087c7a8959bf212fc1c7a2d4b92536d410be). 

Thanks!

File: pkg/connector/connector.go [hunk 5]
```diff
@@ -240,9 +246,18 @@ func validateConfig(awsc *cfg.Aws) error {
 	return nil
 }
 
-func New(ctx context.Context, awsc *cfg.Aws, _ *cli.ConnectorOpts) (connectorbuilder.ConnectorBuilderV2, []connectorbuilder.Opt, error) {
+func New(ctx context.Context, awsc *cfg.Aws, connectorOpts *cli.ConnectorOpts) (connectorbuilder.ConnectorBuilderV2, []connectorbuilder.Opt, error) {
 	l := ctxzap.Extract(ctx)
 
+	// Default to "will sync" when the caller gave no explicit resource-type filter (or no
+	// ConnectorOpts at all), matching ConnectorOpts.WillSyncResourceType's own default.
+	willSyncOrganization := true
```
Comment (sergiocorral-conductorone) [inline pkg/connector/connector.go:256]: **[Major]** I traced this further into how the C1 platform actually drives a real sync, not just the CLI/local path. In `baton-sdk`'s `pkg/tasks/c1api/full_sync.go`, the per-sync-task resource-type filter — the one the platform/UI sends, and which the code's own comment calls "the authoritative source when set" — is read via `c.task.GetSyncFull().GetSyncResourceTypeIds()` and passed **only** to `sdkSync.WithSyncResourceTypes(...)` on the SDK's syncer. It never reaches the connector's own `ConnectorOpts.SyncResourceTypeIDs`, which is fixed once at connector-construction time from local/static config only (`pkg/cli/commands.go`'s `sync-resource-types` flag).

So in service mode, if a C1 tenant opts org/OU out through the platform/UI (rather than through connector-level config), this gate still computes `willSyncOrganization = true` — the syncer will filter org/OU out downstream, but the connector doesn't know that, so it still attaches the account to a parent the syncer will never emit. That's exactly the dangling "MISSING RESOURCE" parent this PR exists to prevent, reachable via the platform's normal opt-in mechanism rather than local CLI flags.

Worth confirming with platform/SDK folks how org/OU opt-in is actually expressed for a real customer tenant — if it's this per-task field, the connector has no way to see it today, and this gate would need either that plumbed through, or the limitation documented plus a tracked SDK-side follow-up.

File: pkg/connector/organization_test.go [hunk 6]
```diff
@@ -264,12 +264,80 @@ func TestAccountList_FlatWhenOrgReadDenied(t *testing.T) {
...
+	require.NoError(t, err)
+	require.Len(t, resources, 1)
+	assert.Nil(t, resources[0].ParentResourceId, "account must stay flat when hierarchy isn't synced")
+	assert.Equal(t, 0, orgs.listParentsCalls, "must not call ListParents when neither hierarchy type is synced")
+}
+
+// account.List resolves the parent but only attaches it when the resolved parent's specific
+// type (organization vs organizational_unit) is one this run will actually sync — partial
+// opt-in must not produce a dangling parent either.
+func TestAccountList_SkipsReparentWhenResolvedParentTypeNotSynced(t *testing.T) {
```
Comment (sergiocorral-conductorone) [inline pkg/connector/organization_test.go:329]: **[Minor]** The specific branch that changed in `8d39453` — org=false, OU=true — has no test. `TestAccountList_SkipsReparentWhenHierarchyNotSynced` covers (false,false) and this test covers (true,false); a third case via `newOrgAccountWithSyncFilter(orgs, false, true)` asserting `orgs.listParentsCalls == 0` would lock in the simplified guard so a future re-widening back to `org || ou` doesn't pass CI silently.

File: pkg/connector/account.go [hunk 7]
```diff
@@ -190,15 +206,29 @@ func (o *accountResourceType) List(ctx context.Context, _ *v2.ResourceId, opts r
...
-			resourceOpts = append(resourceOpts, resourceSdk.WithParentResourceID(parentID))
+		//
+		// Gated on willSyncOrganization: organization and organizational_unit are OptInRequired,
+		// so a sync run may never emit them. Pointing an account's parent at a resource type this
+		// run isn't syncing produces a dangling "MISSING RESOURCE" parent that c1 silently drops
+		// (see permissionSetRoleID's comment on dangling references). The OU crawl is itself
+		// seeded exclusively from Root (see organization.go), so an OU can never actually be
+		// synced when organization is not - skip the ListParents call entirely in that case
+		// rather than resolving a parent that can never be attached.
+		if o.hierarchySync.Organization {
```
Comment (FeliLucero1) [inline pkg/connector/account.go:207]: Nice fix overall — but heads up, this simplification has a side effect nobody's called out yet: in the (org=false, OU=true) case, the old code still fired `ListParents` and could flip `orgReadDenied` → trigger the WARN about missing `organizations:ListParents` permission. Now that whole block is skipped, so that WARN can never fire in this combo either. End result (no parent attached) is the same either way, so no data bug — just a lost diagnostic, and it's the exact combo that's still missing a test (per the open ask from sergiocorral / luisina-santos / the CI bot). Not blocking, just flagging it so it's a documented tradeoff instead of an accidental side effect once that test gets added.

File: pkg/connector/account.go [hunk 8]
```diff
@@ -190,15 +206,29 @@ func (o *accountResourceType) List(ctx context.Context, _ *v2.ResourceId, opts r
...
-		if err != nil {
-			return nil, nil, err
-		}
-		if accessDenied {
-			orgReadDenied = true
-		}
-		if parentID != nil {
-			resourceOpts = append(resourceOpts, resourceSdk.WithParentResourceID(parentID))
+		//
+		// Gated on willSyncOrganization: organization and organizational_unit are OptInRequired,
```
Comment (FeliLucero1) [inline pkg/connector/account.go]: Minor: this comment still says "Gated on willSyncOrganization" but the field's now `hierarchySync.Organization` (see the struct above). Doubly confusing because `connector.go`'s `AWS` struct still has an actual field called `willSyncOrganization` — so grepping that name lands you on the wrong struct entirely. One-line fix, not blocking.

**Review by FeliLucero1 — APPROVED** [review]
Approving — the refactor holds up. Verified the guard simplification is behavior-equivalent for the 3 (Organization, OrganizationalUnit) combos where it matters (attach outcome unchanged), and traced through the OU-crawl seeding logic + the SDK's `WillSyncResourceType` to confirm the org→OU dependency claim in the new comment is accurate. Left 2 small non-blocking notes inline — one's a genuinely new observation building on the test-coverage gap already open on this thread, the other's just a stale comment. Neither should hold up merge.

File: pkg/connector/connector.go (hunk 5 above)
Comment (JavierCarnelli-ConductorOne) [reply pkg/connector/connector.go:256]: I checked the code in C1 and the baton-sdk and can confirm that the concern is not accurate for how the connectors recieve the configs set by the actual C1 users.
The resources configurated to sync via the UI goes through a C1 function which writes the connectors ResourceTypeIds and triggers a redeployment of the connector's Lambda with regenerated static config which is exactly what becomes the ConnectorOpts.SyncResourceTypeIDs at connector construction.

Thanks for the comment tho!

File: pkg/connector/account.go (hunk 8 above)
Comment (JavierCarnelli-ConductorOne) [reply pkg/connector/account.go]: Modified in [6dba262](https://github.com/ConductorOne/baton-aws/pull/139/commits/6dba26292f54b539523f5afdcd28404bd90e13b2)
Thanks!
